### PR TITLE
Apply: models

### LINKED ISSUE #1: Apply: Persisted Models
### New models
- Program: has many sessions, program_questions
  - name: string
  - description: text
- TimeBlock: has many sessions
  - Start date: date
  - End date: date
  - Internal name: string
  - external name: string
- Session: belongs to Program and Block, has many applications
  - Capacity: int
  - status: int (enum: open, closed, full)
- Question: has many responses, program_questions
  - text: string
  - type: int (enum: field, area, number, file)
- ProgramQuestion: belongs to program, question
- Application: belongs to user, session. Validates user and session
  - submitted_at: datetime
  - accepted_at: datetime
  - tentatively_attending_at
  - confirmed_attending_at
  - deactivated_at: datetime
- Response: belongs to user, question
  - type: enum (must match types of question via constants obj)
  - answer: *investigate type*

### Updates to user
 - has many responses
 - has many applications

### LINKED ISSUE #2: Apply: ActiveType Models
- Submission
  - No inherited persistence
  - Validates that all unique questions, as determined by selected programs, have responses. 
  - After create: updates submitted_at on application, creates a submitted event. 
- Active application
  - Inherits from Application
  - validates deactivated_at is not present
- SubmittedApplication _(used in submission model)_
  - inherits from ActiveApplication
  - Validates submitted_at present

### PR DESCRIPTION
**Pull request by JoshBroomberg**
Builds out the models required for the apply section. 

Closes #1 
Closes #2

### CODE REVIEW

File: app/models/journey/application.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class Journey::Application < ActiveType::Record[Journey]
+  validates_presence_of :submitted_at
+
+  def questions
+    self.session.program.program_questions.map(&:question)
```
Comment (khanotations) [inline app/models/journey/application.rb]: Please use a joins, not a map, so we can keep this a relation

File: app/models/session.rb [hunk 2]
```diff
@@ -2,6 +2,8 @@ class Session < ActiveRecord::Base
   belongs_to :program
   belongs_to :period
 
+  validates_presence_of :program_id, :period_id
```
Comment (khanotations) [inline app/models/session.rb]: `program` and `period`, surely, like we do in `Journey`?

File: app/models/submission.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    if (application.questions - application.user.responses.map(&:question)).any?
+      errors.add(:base, "Some questions for the desired program have not been answered.")
+      return false
+    end
+    true
+  end
+
+  def application
+    @application ||= Journey::Application.includes(session: {program: {program_questions: :question}}, user: {responses: :question})
+      .find(self.journey.id)
```
Comment (khanotations) [inline app/models/submission.rb]: Isn't this the same as `self.id`?

Comment (JoshBroomberg) [inline app/models/submission.rb]: No, the submission is a wrapper object with Journey as an `attr_accessor`.

File: app/models/session.rb (hunk 2 above)
Comment (JoshBroomberg) [inline app/models/session.rb]: Fair

File: app/models/journey/application.rb (hunk 1 above)
Comment (JoshBroomberg) [inline app/models/journey/application.rb]: Fair

Comment (JoshBroomberg) [inline app/models/journey/application.rb]: I have done this upstream in #16

File: app/models/session.rb (hunk 2 above)
Comment (JoshBroomberg) [inline app/models/session.rb]: Done upstream in #16

**Comment by JoshBroomberg** [comment]
@khanotations I have made these fixes upstream. Let's merge this to get it out the way and clean up the complex diffs.

File: app/models/journey/application.rb (hunk 1 above)
Comment (khanotations) [inline app/models/journey/application.rb]: I don't see this method in the upstream PR>

Comment (khanotations) [inline app/models/journey/application.rb]: This was not changed in the upstream PR, at least in `journey.rb#questions`.

File: app/models/session.rb (hunk 2 above)
Comment (khanotations) [inline app/models/session.rb]: This change was also not made.

**Pull request merged**